File: problems/0301_remove-invalid-parentheses/solution.cpp
class Solution {
  public:
    vector<string> removeInvalidParentheses(string s) {
        auto isValid = [](const string &str) {
            int count = 0;
            for (char ch : str) {
                if (ch == '(')
                    count++;
                else if (ch == ')') {
                    count--;
                    if (count < 0)
                        return false;
                }
            }
            return count == 0;
        };
        unordered_set<string> level;
        level.insert(s);
        while (true) {
            vector<string> valid;
            for (const string &item : level) {
                if (isValid(item))
                    valid.push_back(item);
            }
            if (!valid.empty()) {
                sort(valid.begin(), valid.end());
                return valid;
            }
            unordered_set<string> next;
            for (const string &item : level) {
                for (int i = 0; i < (int)item.size(); i++) {
                    char ch = item[i];
                    if (ch == '(' || ch == ')') {
                        next.insert(item.substr(0, i) + item.substr(i + 1));
                    }
                }
            }
            level = std::move(next);
        }
    }
};

File: problems/0301_remove-invalid-parentheses/solution.go
import "sort"

func removeInvalidParentheses(s string) []string {
	isValid := func(str string) bool {
		count := 0
		for _, ch := range str {
			if ch == '(' {
				count++
			} else if ch == ')' {
				count--
				if count < 0 {
					return false
				}
			}
		}
		return count == 0
	}
	level := map[string]bool{s: true}
	for {
		valid := []string{}
		for item := range level {
			if isValid(item) {
				valid = append(valid, item)
			}
		}
		if len(valid) > 0 {
			sort.Strings(valid)
			return valid
		}
		next := map[string]bool{}
		for item := range level {
			for i := 0; i < len(item); i++ {
				ch := item[i]
				if ch == '(' || ch == ')' {
					next[item[:i]+item[i+1:]] = true
				}
			}
		}
		level = next
	}
}
